### PR TITLE
Explainer virtual service cannot work

### PR DESCRIPTION
**Pull request by hougangliu**
**What this PR does / why we need it**:
Fixes: #445 
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

```

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kfserving/446)

### CODE REVIEW

File: pkg/controller/inferenceservice/resources/istio/virtualservice.go [hunk 1]
```diff
@@ -84,54 +84,66 @@ func createFailedStatus(reason string, message string) *v1alpha2.VirtualServiceS
...
-		predictCanarySpec, reason := getPredictStatusConfigurationSpec(isvc.Spec.Canary, isvc.Status.Canary)
-		if predictCanarySpec == nil {
-			return nil, createFailedStatus(reason, "Failed to reconcile canary predictor")
+	if endpointSpec.Explainer != nil {
+		explainDefaultSpec, explainerReason := getExplainStatusConfigurationSpec(endpointSpec, endpointStatusMap)
+		if explainDefaultSpec != nil {
+			httpRouteDestination := createHTTPRouteDestination(explainDefaultSpec.Hostname, weight, r.ingressConfig.IngressServiceName)
+			return &httpRouteDestination, nil
+		} else {
+			return nil, createFailedStatus(explainerReason, "Failed to reconcile default explainer")
```
Comment (cliveseldon) [inline pkg/controller/inferenceservice/resources/istio/virtualservice.go]: Is this also called for canary route?

File: pkg/controller/inferenceservice/resources/istio/virtualservice.go [hunk 2]
```diff
@@ -210,6 +224,14 @@ func (r *VirtualServiceBuilder) CreateVirtualService(isvc *v1alpha2.InferenceSer
 	return &vs, &status
 }
 
+func getServiceHostname(endpointSpec *v1alpha2.EndpointSpec, endpointStatusMap *v1alpha2.EndpointStatusMap) (string, error) {
+	predictSpec, reason := getPredictStatusConfigurationSpec(endpointSpec, endpointStatusMap)
+	if predictSpec == nil {
+		return "", fmt.Errorf("Fail to get servie hostname: %s.", reason)
```
Comment (cliveseldon) [inline pkg/controller/inferenceservice/resources/istio/virtualservice.go]: typo "servie" -> "service"

Comment (hougangliu) [reply pkg/controller/inferenceservice/resources/istio/virtualservice.go]: done! thanks

File: pkg/controller/inferenceservice/resources/istio/virtualservice.go (hunk 1 above)
Comment (hougangliu) [reply pkg/controller/inferenceservice/resources/istio/virtualservice.go]: yes, I need refactor the var name

**Comment by rakelkar** [comment]
@hougangliu you might also want to look at https://github.com/kubeflow/kfserving/pull/434 which fixes this bug and also places the transformer in front of the explainer...

**Comment by rakelkar** [comment]
Your refactor seems very reasonable, could you also add a test for virtualservice.go?

**Comment by ellis-bigelow** [comment]
Haven't dived into the code. Does this change put explainer -> transformer or transformer -> explainer?

**Comment by hougangliu** [comment]
@ellis-bigelow any reason why close this PR?

**Comment by ellis-bigelow** [comment]
Misclicked!

**Comment by hougangliu** [comment]
> Haven't dived into the code. Does this change put explainer -> transformer or transformer -> explainer?

This PR is used to fix kfserving-controller-manager-0 crash issue when we run a InferenceService with explainer. for explainer and transformer, this PR remains current behavior

**Comment by hougangliu** [comment]
> Your refactor seems very reasonable, could you also add a test for virtualservice.go?

I tried, but it always failed. How about adding a test in next PR in which places the transformer in front of the explainer?

**Comment by ellis-bigelow** [comment]
> > Haven't dived into the code. Does this change put explainer -> transformer or transformer -> explainer?
> 
> This PR aims to fix kfserving-controller-manager-0 crash issue when we run a InferenceService with explainer. for relation between explainer and transformer, this PR remains current behavior.

If possible, given that we know there's already a bug in here, and it's already being refactored, I think it would be great to just do the changes together. If you think this is overly complex, I'm fine submitting this as is, but I think it generates extra work.

**Comment by hougangliu** [comment]
> If possible, given that we know there's already a bug in here, and it's already being refactored, I think it would be great to just do the changes together. If you think this is overly complex, I'm fine submitting this as is, but I think it generates a little extra work.

In fact, it will not generates any extra work. #434 can drop its change in  pkg/controller/inferenceservice/resources/istio/virtualservice.go and only need add below change based on this PR. @rakelkar make sense?
```
        if endpointSpec.Explainer != nil {
                explainSpec, explainerReason := getExplainStatusConfigurationSpec(endpointSpec, endpointStatusMap)
                if explainSpec != nil {
-                       httpRouteDestination := createHTTPRouteDestination(explainSpec.Hostname, weight, r.ingressConfig.IngressServiceName)
+                       hostname := explainSpec.Hostname
+                       if endpointSpec.Transformer != nil {
+                               transformerSpec, reason := getTransformerStatusConfigurationSpec(endpointSpec, endpointStatusMap)
+                               if transformerSpec == nil {
+                                       return nil, createFailedStatus(reason, "Failed to reconcile transformer")
+                               }
+                               hostname = transformerSpec.Hostname
+                       }
+
+                       httpRouteDestination := createHTTPRouteDestination(hostname, weight, r.ingressConfig.IngressServiceName)
                        return &httpRouteDestination, nil
```

**Comment by ellis-bigelow** [comment]
Sgtm

**Comment by rakelkar** [comment]
Yes makes sense - thanks for this fix!
/approve
/lgtm

**Pull request merged**